### PR TITLE
add --repoid/--repo and --repofrompath options

### LINKED ISSUE #234: Implement the --repoid option
Using `--disablerepo=* --enablerepo=some-rep` is tedious. The `--repoid some-repo` would make it easier to select a particular repo.

### LINKED ISSUE #235: Implement the repofrompath option
The `--repofrompath` option would make it easy to use an ad-hoc repository, for example one just created with the `reposync` command.

### PR DESCRIPTION
**Pull request by oliverkurth**
This addresses issues #234 and #235 . Usages:
```
tdnf --repo=<repoid> ...
tdnd --repofrompath=<repoid>,<url> --repo=<repoid> ...
```
Tests are added as well.

### CODE REVIEW

File: client/defines.h [hunk 1]
```diff
@@ -46,6 +49,8 @@ typedef enum
         }                                                          \
     } while(0)
 
+#define STR_IS_TRUE(s) (s && (!strcmp(s, "1") || !strcasecmp(s, "true")))
```
Comment (sshedi) [inline client/defines.h]: ignore case while checking for true.
use braces for `s`, we might pass an address with some pointer math

File: client/defines.h [hunk 2]
```diff
@@ -28,6 +28,9 @@ typedef enum
     DETAIL_INFO
 }TDNF_PKG_DETAIL;
 
+#define STRINGIFYX(N) STRINGIFY(N)
```
Comment (sshedi) [inline client/defines.h]: Not sure if feasible, can this be moved to some common place? So that other modules can also use it if needed.

File: client/defines.h [hunk 3]
```diff
@@ -131,7 +136,8 @@ typedef enum
 #define TDNF_REPO_DEFAULT_SSLVERIFY       1
 #define TDNF_REPO_DEFAULT_RETRIES         10
 #define TDNF_REPO_DEFAULT_PRIORITY        50
-#define TDNF_REPO_DEFAULT_METADATA_EXPIRE "172800" // 48 hours in seconds
+#define TDNF_REPO_DEFAULT_METADATA_EXPIRE 172800 // 48 hours in seconds
+#define TDNF_REPO_DEFAULT_METADATA_EXPIRE_STR STRINGIFYX(TDNF_REPO_DEFAULT_METADATA_EXPIRE)
```
Comment (sshedi) [inline client/defines.h:142]: good idea :)

File: client/repolist.c [hunk 4]
```diff
@@ -49,6 +51,27 @@ TDNFLoadRepoData(
...
     BAIL_ON_TDNF_ERROR(dwError);
 
+    for(pSetOpt = pTdnf->pArgs->pSetOpt;
+        pSetOpt;
+        pSetOpt = pSetOpt->pNext)
+    {
+        if(strcmp(pSetOpt->pszOptName, "repofrompath") == 0)
+        {
+            TDNFSplitStringToArray(pSetOpt->pszOptValue, ",", &ppszUrlIdTuple);
+            if ((ppszUrlIdTuple[0] == NULL) || ppszUrlIdTuple[1] == NULL)
```
Comment (sshedi) [inline client/repolist.c:61]: Can you give an example of how inputs comes here? Will there be minimum 2 items in the ppszUrlIdTuple array always?

Can we give multiple repo paths at the same time?

File: client/repolist.c [hunk 5]
```diff
@@ -49,6 +51,27 @@ TDNFLoadRepoData(
...
+            {
+                dwError = ERROR_TDNF_INVALID_PARAMETER;
+                BAIL_ON_TDNF_ERROR(dwError);
+            }
+            dwError = TDNFCreateRepoFromPath(&pReposAll,
+                ppszUrlIdTuple[0], ppszUrlIdTuple[1]);
```
Comment (sshedi) [inline client/repolist.c]: nit: indent it properly

File: client/repolist.c [hunk 6]
```diff
@@ -111,6 +134,7 @@ TDNFLoadRepoData(
         closedir(pDir);
     }
     TDNF_SAFE_FREE_MEMORY(pszRepoFilePath);
+    TDNF_SAFE_FREE_STRINGARRAY(ppszUrlIdTuple);
```
Comment (sshedi) [inline client/repolist.c:138]: why is this needed? if ppszUrlIdTuple[1] is null and ppszUrlIdTuple[0] has garbage address, it might crash.

File: client/repolist.c [hunk 7]
```diff
@@ -139,17 +163,124 @@ TDNFCreateCmdLineRepo(
         BAIL_ON_TDNF_ERROR(dwError);
     }
 
+    dwError = TDNFCreateRepo(&pRepo, "@cmdline");
```
Comment (sshedi) [inline client/repolist.c]: Please move `@cmdline` to a macro, it's usage is spread across files now.

File: client/repolist.c [hunk 8]
```diff
@@ -139,17 +163,124 @@ TDNFCreateCmdLineRepo(
...
+    {
+        TDNFFreeReposInternal(pRepo);
+    }
+    goto cleanup;
+}
+
+uint32_t
+TDNFCreateRepoFromPath(
+    PTDNF_REPO_DATA_INTERNAL* ppRepo,
+    const char *pszId,
```
Comment (sshedi) [inline client/repolist.c:187]: validate function args

File: client/repolist.c [hunk 9]
```diff
@@ -139,17 +163,124 @@ TDNFCreateCmdLineRepo(
...
+    BAIL_ON_TDNF_ERROR(dwError);
+
+    dwError = TDNFSafeAllocateString("@cmdline", &pRepo->pszName);
+    BAIL_ON_TDNF_ERROR(dwError);
+
+    *ppRepo = pRepo;
+cleanup:
+    return dwError;
+error:
+    if(ppRepo)
```
Comment (sshedi) [inline client/repolist.c]: No need to set ppRepo to NULL

File: pytests/tests/test_repofrompath.py [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+                     'makecache'],
+                    cwd=workdir)
+    assert(ret['retval'] == 0)
+
+    pkgname = utils.config["mulversion_pkgname"]
+    utils.erase_package(pkgname)
+    ret = utils.run(['tdnf',
+                     '-y', '--nogpgcheck',
+                     '--repofrompath=synced-repo,{}'.format(synced_dir),
+                     '--repo=synced-repo',
```
Comment (sshedi) [inline pytests/tests/test_repofrompath.py:64]: Why do we need to specify `--repo` again? it's already given with `--repofrompath`
it will be good if we can specify multiple repos in one shot.
Like: `tdnf -y --repofrompath=repo-a,<repo-a-path>,repo-b,<repo-b-path>`

If what you have done is in alignment with dnf, let's use the same. This is just an idea

**Review by sshedi — CHANGES_REQUESTED** [review]
Thanks for the feature. Minor changes needed.

File: client/defines.h (hunk 1 above)
Comment (oliverkurth) [reply client/defines.h]: The comparsion is case insensitive already - `strcasecmp()`.

Braces added, thanks!

File: client/defines.h (hunk 2 above)
Comment (oliverkurth) [reply client/defines.h]: Moved to `common/defines.h`.

File: client/repolist.c (hunk 4 above)
Comment (oliverkurth) [reply client/repolist.c:61]: If the syntax is correctly given then `ppszUrlIdTuple` will have exactly two items: the repoid and the URL.

You can specify multiple paths by giving the option multiple times.

File: client/repolist.c (hunk 6 above)
Comment (oliverkurth) [reply client/repolist.c:138]: How would `ppszUrlIdTuple[0]` have garbage? It's allocated and initialized in `TDNFSplitStringToArray()`. This is a pretty common pattern throughout the code.

File: pytests/tests/test_repofrompath.py (hunk 10 above)
Comment (oliverkurth) [reply pytests/tests/test_repofrompath.py:64]: Setting `--repo` disables all repos and enable only this one, which is what we want. It's a shortcut for `--disablerepo="*" --enablerepo=<repoid>`.

As explained in another comment, you can specify multiple repos with multiple `--repo` options.

File: client/repolist.c (hunk 7 above)
Comment (oliverkurth) [reply client/repolist.c]: > Thanks for the feature. Minor changes needed.

Thanks for the review - I totally missed your comments for almost 3 weeks 😞 .

File: client/repolist.c [hunk 11]
```diff
@@ -139,17 +164,120 @@ TDNFCreateCmdLineRepo(
...
+}
+
+uint32_t
+TDNFCreateRepoFromPath(
+    PTDNF_REPO_DATA_INTERNAL* ppRepo,
+    const char *pszId,
+    const char *pszPath
+    )
+{
+    uint32_t dwError;
```
Comment (tapakund) [inline client/repolist.c]: better to initialise dwError = 0 in all the function added.

**Review by tapakund — APPROVED** [review]
Looks good.
